lib_static/plugins/acts_as_journalized/lib/cause_of_change/base.rb

```diff
@@ -0,0 +1,56 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module CauseOfChange
+  class Base
+    attr_reader :type, :additional_attributes
+
+    def initialize(type, additional_attributes = {})
+      @type = type
+      @additional_attributes = additional_attributes
+    end
+
+    def to_h
+      { "type" => type }.merge(additional_attributes).deep_stringify_keys
+    end
+
+    def blank?
+      false
+    end
+
+    def ==(other)
+      if other.is_a?(Hash)
+        to_h == other
+      else
+        other.class == self.class &&
+         type == other.type &&
+         additional_attributes == other.additional_attributes
+      end
+    end
```

Comment:
Implementing the `==` method with the `to_hash` call would also yield the benefit of explicit conversion. Hence we  can omit explicitly calling `to_h` in case of comparison with another hash.
The calls such as `predecessor.cause == cause.to_h` could be just simply written as `predecessor.cause == cause`

```suggestion
     def ==(other)
      other.to_hash == to_hash
    end
```

This is just a convenience suggestion.